File: electron/agentToolDefinitions.js
/**
 * Agent 工具定义（编排元数据与格式转换）
 *
 * - 发给大模型的 `description` / `parameters`：**仅**来自 `electron/skills/<name>/SKILL.md`（由 `toolRouter` 扫描目录加载），**不再**使用本文件的重复清单。
 * - 本文件保留 **`SKILL_SPECS`**（DAG / 编排）与 **`toOpenAiTools`**；新增工具：在 `toolExecutor` 实现 + 增加 `SKILL.md` + 按需补 `SKILL_SPECS`。
 */

/**
 * @typedef {{
 *   name: string,
 *   description: string,
 *   parameters: object,
 *   skillSpec?: SkillSpec
 * }} RouterSkillItem
 */
/**
 * @typedef {{
 *   requires?: string[],
 *   provides?: string[],
 *   consumes?: string[],
 *   riskLevel: 'read'|'write',
 *   autoResolveArgs?: string[]
 * }} SkillSpec
 */

/**
 * OpenAI 兼容的 tools 数组（完整 function 声明）
 * @param {RouterSkillItem[]} items
 */
function toOpenAiTools(items) {
  return items.map((s) => ({
    type: 'function',
    function: {
      name: s.name,
      description: s.description || '',
      parameters: s.parameters,
    },
  }))
}

/**
 * 技能结构化元信息（DAG 规划/执行器使用）。
 * - 未显式配置的技能：默认 read + 无依赖。
 * @type {Record<string, SkillSpec>}
 */
const SKILL_SPECS = {
  getGlobalOutline: {
    riskLevel: 'read',
    provides: ['globalOutlineMarkdown'],
    consumes: ['bookId'],
  },
  editGlobalOutline: {
    riskLevel: 'write',
    requires: ['getGlobalOutline'],
    consumes: ['bookId', 'markdownContent'],
    autoResolveArgs: ['bookId'],
  },
  listOutlines: {
    riskLevel: 'read',
    provides: ['outlinesIndex'],
    consumes: ['bookId'],
  },
  queryOutline: {
    riskLevel: 'read',
    requires: ['listOutlines'],
    provides: ['outlineDetails'],
    consumes: ['bookId'],
    autoResolveArgs: ['bookId', 'outlineIds'],
  },
  updateOutline: {
    riskLevel: 'write',
    requires: ['listOutlines', 'queryOutline'],
    consumes: ['bookId', 'outlineId'],
    autoResolveArgs: ['bookId', 'outlineId'],
  },
  listWritingChapters: {
    riskLevel: 'read',
    provides: ['writingChaptersIndex'],
    consumes: ['bookId'],
  },
  getChapterContent: {
    riskLevel: 'read',
    requires: ['listWritingChapters'],
    consumes: ['bookId', 'chapterId'],
  },
  batchGetChapterContents: {
    riskLevel: 'read',
    requires: ['listWritingChapters'],
    consumes: ['chapterIds'],
  },
  editChapterContent: {
    riskLevel: 'write',
    requires: ['listWritingChapters'],
    consumes: ['chapterId', 'content'],
    autoResolveArgs: ['chapterId'],
  },
  addForeshadowing: {
    riskLevel: 'write',
    requires: ['listWritingChapters'],
    consumes: ['bookId', 'chapterId', 'content'],
  },
}

/**
 * 与 `toolRouter.getApiSkillItems()` 中的工具名对齐；须在 main 已 `setSkillsPath` 之后调用。
 * @returns {Record<string, SkillSpec>}
 */
function getSkillSpecs() {
  const toolRouter = require('./toolRouter')
  const items = toolRouter.getApiSkillItems()
  const out = {}
  for (const item of items) {
    const raw = SKILL_SPECS[item.name] || {}
    out[item.name] = {
      requires: Array.isArray(raw.requires) ? raw.requires : [],
      provides: Array.isArray(raw.provides) ? raw.provides : [],
      consumes: Array.isArray(raw.consumes) ? raw.consumes : [],
      riskLevel: raw.riskLevel === 'write' ? 'write' : 'read',
      autoResolveArgs: Array.isArray(raw.autoResolveArgs) ? raw.autoResolveArgs : [],
    }
  }
  return out
}

module.exports = {
  SKILL_SPECS,
  getSkillSpecs,
  toOpenAiTools,
}

File: backend/services/agent_tool_definitions.py
"""
Agent tool definitions — port of electron/agentToolDefinitions.js.

SKILL_SPECS (DAG / orchestration metadata) and format conversion to
OpenAI-compatible tools arrays.
"""

from __future__ import annotations

from typing import Any


# ---------------------------------------------------------------------------
# Skill specs (DAG dependencies, read/write risk)
# ---------------------------------------------------------------------------

SKILL_SPECS: dict[str, dict[str, Any]] = {
    "getGlobalOutline": {
        "riskLevel": "read",
        "provides": ["globalOutlineMarkdown"],
        "consumes": ["bookId"],
    },
    "editGlobalOutline": {
        "riskLevel": "write",
        "requires": ["getGlobalOutline"],
        "consumes": ["bookId", "markdownContent"],
        "autoResolveArgs": ["bookId"],
    },
    "listOutlines": {
        "riskLevel": "read",
        "provides": ["outlinesIndex"],
        "consumes": ["bookId"],
    },
    "queryOutline": {
        "riskLevel": "read",
        "requires": ["listOutlines"],
        "provides": ["outlineDetails"],
        "consumes": ["bookId"],
        "autoResolveArgs": ["bookId", "outlineIds"],
    },
    "updateOutline": {
        "riskLevel": "write",
        "requires": ["listOutlines", "queryOutline"],
        "consumes": ["bookId", "outlineId"],
        "autoResolveArgs": ["bookId", "outlineId"],
    },
    "listWritingChapters": {
        "riskLevel": "read",
        "provides": ["writingChaptersIndex"],
        "consumes": ["bookId"],
    },
    "createWritingChapter": {
        "riskLevel": "write",
        "requires": ["listWritingChapters"],
        "consumes": ["bookId"],
        "autoResolveArgs": ["bookId"],
    },
    "getChapterContent": {
        "riskLevel": "read",
        "requires": ["listWritingChapters"],
        "consumes": ["bookId", "chapterId"],
    },
    "batchGetChapterContents": {
        "riskLevel": "read",
        "requires": ["listWritingChapters"],
        "consumes": ["chapterIds"],
    },
    "editChapterContent": {
        "riskLevel": "write",
        "requires": ["listWritingChapters"],
        "consumes": ["chapterId", "content"],
        "autoResolveArgs": ["chapterId"],
    },
    "addForeshadowing": {
        "riskLevel": "write",
        "requires": ["listWritingChapters"],
        "consumes": ["bookId", "chapterId", "content"],
    },
    "listBookCharacters": {
        "riskLevel": "read",
        "provides": ["bookCharactersIndex"],
        "consumes": ["bookId"],
    },
    "getBookCharacters": {
        "riskLevel": "read",
        "requires": ["listBookCharacters"],
        "provides": ["bookCharactersDetails"],
        "consumes": ["bookId"],
    },
    "createCharacter": {
        "riskLevel": "write",
        "requires": ["listBookCharacters"],
        "consumes": ["bookId", "name"],
        "autoResolveArgs": ["bookId"],
    },
    "updateCharacter": {
        "riskLevel": "write",
        "requires": ["listBookCharacters"],
        "consumes": ["bookId", "characterId"],
        "autoResolveArgs": ["bookId"],
    },
    "editStoryBackground": {
        "riskLevel": "write",
        "requires": ["getStoryBackground"],
        "consumes": ["bookId", "content"],
        "autoResolveArgs": ["bookId"],
    },
}


# ---------------------------------------------------------------------------
# OpenAI-compatible tools conversion
# ---------------------------------------------------------------------------

def to_openai_tools(items: list[dict]) -> list[dict]:
    """Convert skill items (from tool_router) into OpenAI function-call tool array."""
    return [
        {
            "type": "function",
            "function": {
                "name": s["name"],
                "description": s.get("description", ""),
                "parameters": s.get("parameters", {"type": "object", "properties": {}, "required": []}),
            },
        }
        for s in items
    ]


def get_skill_specs() -> dict[str, dict[str, Any]]:
    """Return skill specs aligned with the currently loaded tool names.

    Must be called after ``tool_router.set_skills_path()`` has been invoked.
    """
    from services.tool_router import get_api_skill_items

    items = get_api_skill_items()
    out: dict[str, dict] = {}
    for item in items:
        raw = SKILL_SPECS.get(item["name"], {})
        out[item["name"]] = {
            "requires": list(raw.get("requires", [])),
            "provides": list(raw.get("provides", [])),
            "consumes": list(raw.get("consumes", [])),
            "riskLevel": "write" if raw.get("riskLevel") == "write" else "read",
            "autoResolveArgs": list(raw.get("autoResolveArgs", [])),
        }
    return out
